Stock_Alert/response.py

```diff
@@ -0,0 +1,22 @@
+import requests
+
+
+class Response:
+    def __init__(self, endpoint):
+        self.endpoint = endpoint
+
+    def get_hold_of(self, **kwargs):
+        try:
+            response = requests.get(self.endpoint, params=kwargs)
+            response.raise_for_status()
+            return response
+        except ValueError as v:
```

Comment:
![Stock_alert_weird error](https://github.com/sasha-cz/stock_news/assets/139366657/fa73c82c-8857-4902-bec8-0aabe99d11ed)
Hm, very strange thing here: 
If a wrong API key is passed in to 'stock_api_key', it still shows the 'stock_data' without error message when printed. Here the program proceeds instead of stopping as you said. When I pass a wrong API key to 'news_api_key', the program does what it should: the last exception 'requests.exceptions.RequestException' gets triggered, the error message is shown and the program exits. I still don't know what exactly the root cause is and how my code is involved in that problem. So I'm a bit stucked here. Do you have an idea how to tackle this? Is the root cause for sure in the Response class (actualized to 'GetRequest')?

Reply:
I also refactored the code by using the params as before and without the Response Class, then I passed in again the wrong key APIs, still I get a response from the stock API. I think I give up for today :)

Reply:
I see, the stock API returns 200 (which is not a standard behavior, but probably their design choice).
![Снимок экрана 2023-11-05 в 17 40 36](https://github.com/sasha-cz/stock_news/assets/19638522/dccbf080-d267-4c58-80d5-8ca0033da85d).

In this case, you can try to get the "Time Series (Daily)" from the response, but handle the case when it's missing with some meaningful error message e.g. 'Time Series (Daily) is missing in the response" + concat it with the error message from the response content. I'd also leave a comment in the code explaining this workaround.

Reply:
I still don't have the perfect solution here: I edited the error handling to catch the key "Error Message" from response.content when status code is still 200 even with a wrong API key. I added also a comment to explain this workaround. But when I run the program locally, I still get hold of the 'Time Series(Daily)' and stock_response.content doesn't contain any error. But it seems it did in your case... So it should catch it as long as the response.content includes the Error Message.

Reply:
My problem in short:  I still get the Time Series (Daily) response running the program on my machine with wrong API key.

Reply:
yes, that's the point of how their API is designed, you get `200` but with an error message in the response. This is a weird choice but it is how it is, we cannot change a third-party API. 

That's why I suggest you don't try to tackle this edge case in this class, but add a check in a place where you get the response - see if `Error Message` is present in the response, then raise an error, e.g.in the main.py:
```
if 'Error Message' in stock_response:
  raise KeyError(stock_response['Error Message'])
else:
  stock_data = stock_response["Time Series (Daily)"]
```

If it's not present, then all good, we proceed with execution.

Reply:
A KeyError would be raised when trying to access a key in a dictionary that does not exist. But here I want to raise an exception if the key "Error Message" exists in the response. That's why I changed your suggestion to the following work around in main.py:

`if 'Error Message' in stock_response.json():`
    `raise Exception(f"{stock_response.json()['Error Message']}")`
`else:`
    `stock_data = stock_response.json()["Time Series (Daily)"]`

Reply:
raising Exception is too general, when we raise an error on purpose like in this case, we tend to give it a more specific name. In your case the key "Time Series (Daily)" doesn't exist, that's why I suggested a KeyError, but also feel free to choose a different error if you feel it suits you better

Reply:
ah ok, I missunderstood it then